openwisp_utils/admin_theme/menu.py

```diff
@@ -0,0 +1,240 @@
+from django.apps import registry
+from django.core.exceptions import ImproperlyConfigured
+from django.urls import reverse
+from django.utils.translation import gettext as _
+
+from ..utils import SortedOrderedDict
+
+MENU = SortedOrderedDict()
+
+
+class BaseMenuItem:
+    '''
+    It is a base class for all types of menu items.
+    It is used to handle some common functions.
+    '''
+
+    def __init__(self, config):
+        if not isinstance(config, dict):
+            raise ImproperlyConfigured(
+                f'"config" should be a type of "dict". Error for config- {config}'
+            )
+
+    def get_context(self, request=None):
+        return self.create_context(request)
+
+    def create_context(self, request=None):
+        return {'label': self.label, 'url': self.url, 'icon': self.icon}
+
+    def set_label(self, config):
+        label = config.get('label')
+        if not label:
+            raise ImproperlyConfigured(f'"label" is missing in the config- {config}')
+        if not isinstance(label, str):
+            raise ImproperlyConfigured(
+                f'"level" should be a type of "str". Error for config- {config}'
+            )
+        self.label = _(label)
+
+
+class ModelLink(BaseMenuItem):
+    '''
+    It is to used create a link for a model, like "list view" and "add view".
+    Parameters for the config: name, model, label, icon
+    '''
+
+    def __init__(self, config):
+        super().__init__(config)
+        name = config.get('name')
+        model = config.get('model')
+        if name:
+            if not isinstance(name, str):
+                raise ImproperlyConfigured(
+                    f'"name" should be a type of "str". Error for config-{config}'
+                )
+            self.name = name
+        else:
+            raise ImproperlyConfigured(f'"name" is missing in the config-{config}')
+        if not model:
+            raise ImproperlyConfigured(f'"model" is missing in config-{config}')
+        if not isinstance(model, str):
+            raise ImproperlyConfigured(
+                f'"model" should be a type of "str". Error for config-{config}'
+            )
+        self.model = model
+        self.label = _(self.get_label(config))
+        self.icon = config.get('icon')
+
+    def get_label(self, config=None):
+        label = config.get('label')
+        if label:
+            return label
+        app_label, model = config['model'].split('.')
+        model_class = registry.apps.get_model(app_label, model)
+        return f'{model_class._meta.verbose_name_plural} {self.name}'
+
+    def create_context(self, request):
+        app_label, model = self.model.split('.')
+        model_label = model.lower()
+        url = reverse(f'admin:{app_label}_{model_label}_{self.name}')
```

Comment:
@nemesisdesign , @atb00ker and @kkreitmair I found an issue here. It will give an error at runtime if reverse is not found. We can't check reverse before loading of url config. 
I think we should write a try statement here and should remove the item from the menu if reverse is not found.
What do you think?

Reply:
Hmmm, this can only happen when the app starts for the first time, correct? So let's let the admin see the error so that they can correct their configuration.
It will never be actually visible to the end-user so it shouldn't be a problem as far as I can think...